### PR TITLE
[FrameworkBundle] Fix freshness checks with boolean parameters on routes

### LINKED ISSUE #40017: [FrameworkBundle][Routing] Boolean container parameters for routes trigger cache refresh on each request
**Symfony version(s) affected**: 4.4

**Description**  
Using boolean container parameters for routes triggers a cache refresh on each request.
The bug was introduced with https://github.com/symfony/symfony/pull/30379 . The problem is that bool values get [cast into a string](https://github.com/symfony/symfony/blob/8533ea223e99c8e0161fdf5b32e99e5d7567fb88/src/Symfony/Bundle/FrameworkBundle/Routing/Router.php#L182) and are stored in the meta file as `"0"` or `"1"`. When the `ContainerParametersResourceChecker` checks whether the cache of fresh [it compares a string to a bool](https://github.com/symfony/symfony/blob/8533ea223e99c8e0161fdf5b32e99e5d7567fb88/src/Symfony/Component/DependencyInjection/Config/ContainerParametersResourceChecker.php#L45) (`"1" === true`) which is always false.

**How to reproduce**  
```yaml
# routes.yaml
foo:
    path: /foo
    condition: '%parameter.condition%'
```

A simple test case:
```php
use Symfony\Bundle\FrameworkBundle\Routing\Router;
use Symfony\Component\Config\FileLocator;
use Symfony\Component\Config\ResourceCheckerConfigCache;
use Symfony\Component\Config\ResourceCheckerConfigCacheFactory;
use Symfony\Component\DependencyInjection\Config\ContainerParametersResourceChecker;
use Symfony\Component\DependencyInjection\Container;
use Symfony\Component\Routing\Loader\YamlFileLoader;

$cacheDir = __DIR__.'/cache';

$container = new Container();
$container->set('routing.loader', new YamlFileLoader(new FileLocator(__DIR__.'/config')));

$container->setParameter('parameter.condition', true);

$router = new Router($container, 'routes.yaml', [
    'debug' => true,
    'cache_dir' => $cacheDir,
]);

$resourceCheckers = [
    new ContainerParametersResourceChecker($container),
];

$router->setConfigCacheFactory(new ResourceCheckerConfigCacheFactory($resourceCheckers));

$router->getMatcher(); // trigger cache build

$cache = new ResourceCheckerConfigCache($cacheDir.\DIRECTORY_SEPARATOR.'UrlMatcher.php', $resourceCheckers);

var_dump($cache->isFresh()); // should be true, is false
```

### PR DESCRIPTION
**Pull request by HypeMC**
| Q             | A
| ------------- | ---
| Branch?       | 4.4
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Tickets       | Fix #40017
| License       | MIT
| Doc PR        | -

### CODE REVIEW

**Review by rheh — APPROVED** [review]

File: src/Symfony/Bundle/FrameworkBundle/Routing/Router.php [hunk 1]
```diff
@@ -172,18 +172,25 @@ private function resolve($value)
...
             $resolved = ($this->paramFetcher)($match[1]);
 
             if (\is_bool($resolved)) {
+                $this->collectedParameters[$match[1]] = $resolved;
                 $resolved = (string) (int) $resolved;
             }
 
             if (\is_string($resolved) || is_numeric($resolved)) {
-                $this->collectedParameters[$match[1]] = $resolved;
+                if (!isset($this->collectedParameters[$match[1]])) {
```
Comment (nicolas-grekas) [inline src/Symfony/Bundle/FrameworkBundle/Routing/Router.php:191]: this check is not needed, isn't it?

File: src/Symfony/Bundle/FrameworkBundle/Routing/Router.php [hunk 2]
```diff
@@ -172,18 +172,25 @@ private function resolve($value)
...
+            }
+
             if (preg_match('/^env\((?:\w++:)*+\w++\)$/', $match[1])) {
                 throw new RuntimeException(sprintf('Using "%%%s%%" is not allowed in routing configuration.', $match[1]));
             }
 
             $resolved = ($this->paramFetcher)($match[1]);
 
             if (\is_bool($resolved)) {
+                $this->collectedParameters[$match[1]] = $resolved;
```
Comment (nicolas-grekas) [inline src/Symfony/Bundle/FrameworkBundle/Routing/Router.php:186]: this should be move after the very next line

File: src/Symfony/Bundle/FrameworkBundle/Routing/Router.php (hunk 1 above)
Comment (nicolas-grekas) [inline src/Symfony/Bundle/FrameworkBundle/Routing/Router.php]: Got it. Instead of this check, shouldn't we turn the previous "if" into an "elseif"?

Comment (HypeMC) [reply src/Symfony/Bundle/FrameworkBundle/Routing/Router.php]: @nicolas-grekas We don't have `$resolved` at that point yet. We could move the entire if statement after `$resolved = ($this->paramFetcher)($match[1]);` but that would partially defeat the purpose as the first if statement should prevent any unnecessary code execution.

Comment (HypeMC) [reply src/Symfony/Bundle/FrameworkBundle/Routing/Router.php]: Another possibility that would work:

```php
if (\is_string($resolved) || is_numeric($resolved) || \is_bool($resolved)) { // or \is_scalar($resolved)
    $this->collectedParameters[$match[1]] = $resolved;

    if (\is_bool($resolved)) {
        $resolved = (string) (int) $resolved;
    }

    return (string) $this->resolve($resolved);
}
```

Comment (nicolas-grekas) [reply src/Symfony/Bundle/FrameworkBundle/Routing/Router.php]: I updated the code, thanks :)

**Review by nicolas-grekas — APPROVED** [review]

**Comment by nicolas-grekas** [comment]
Thank you @HypeMC.

**Pull request merged**